cmd/bucket-replication.go

```diff
@@ -117,6 +117,142 @@ func mustReplicater(ctx context.Context, r *http.Request, bucket, object string,
 	return cfg.Replicate(opts)
 }
 
+// returns true if any of the objects being deleted qualifies for replication.
+func mustReplicateDeletes(ctx context.Context, bucket string, objects []ObjectToDelete) bool {
+	c, err := getReplicationConfig(ctx, bucket)
+	if err != nil || c == nil {
+		return false
+	}
+	for _, obj := range objects {
+		if c.HasActiveRules(obj.ObjectName, true) {
+			return true
+		}
+	}
+	return false
+}
+
+// returns true if object qualifies for replication
+func isReplicateDelete(ctx context.Context, getObjectInfoFn GetObjectInfoFn, bucket string, dobj ObjectToDelete) bool {
+	rcfg, err := getReplicationConfig(ctx, bucket)
+	if err != nil || rcfg == nil {
+		return false
+	}
+	oi, err := getObjectInfoFn(ctx, bucket, dobj.ObjectName, ObjectOptions{VersionID: dobj.VersionID})
+	if err != nil {
+		validReplStatus := false
+		switch oi.ReplicationStatus {
+		case replication.Pending, replication.Complete, replication.Failed:
+			validReplStatus = true
+		}
+		if oi.DeleteMarker && validReplStatus {
+			return true
+		}
+		return false
+	}
+	opts := replication.ObjectOpts{
+		Name:         dobj.ObjectName,
+		SSEC:         crypto.SSEC.IsEncrypted(oi.UserDefined),
+		UserTags:     oi.UserTags,
+		DeleteMarker: true,
```

Comment:
shouldn't we be checking `oi.DeleteMarker`, `oi.ReplicationStatus`, etc when `err == nil` too?

Reply:
when getObjectInfo returns not found, it could mean one of two things - either it is a non existent object or a delete marker. We are using the additional metadata returned by erasure layer to decide if versioned delete of deletemarker should be replicated. When err is nil, the object exists - so the replication configuration needs to be checked to see if delete replication is enabled and other rules qualify